### PR TITLE
Update the negotiation and Locale logic to match latest algo

### PR DESCRIPTION
**Pull request by zbraniecki**
This patch brings the JS implementation very close to the C++ and Rust implementations.

### CODE REVIEW

**Comment by zbraniecki** [comment]
it removes the `*` as an option for language tag, because we really never need it. Instead it uses ranges on `match` method which is what we do in C++ and Rust.

**Comment by zbraniecki** [comment]
Fixes #334

File: fluent-langneg/src/locale.js [hunk 1]
```diff
@@ -34,53 +34,63 @@ export default class Locale {
...
-  matches(locale) {
+  matches(locale, thisRange = false, otherRange = false) {
     return localeParts.every(part => {
-      return this[part] === "*" || locale[part] === "*" ||
-        (this[part] === undefined && locale[part] === undefined) ||
-        (this[part] !== undefined && locale[part] !== undefined &&
-        this[part].toLowerCase() === locale[part].toLowerCase());
+      return ((thisRange && this[part] === undefined) ||
+              (otherRange && locale[part] === undefined) ||
+              this[part] === locale[part]);
```
Comment (stasm) [inline fluent-langneg/src/locale.js]: Understanding this method is important for understanding the whole algorithm and I'm not fond of adding two boolean arguments to it. It's hard to follow what they mean when the method is called. In fact, I think I largely preferred the old code which used an explicit `*` to represent ranges. 

A few suggestions on how to improve this code. I'm not saying all of them need to be applied at the same time, but I'd like to see this method improved _somehow_.

- Remove the `thisRange` param as it is always `true` across all callsites, or start using it for exact matching.
- Use an `options` dictionary: `locale.matches(other, {thisRange: true, otherRange: true})`.
- Separate this code into multiple methods: `exactMatches`, `matchesThisRange`, `matchesRanges`.
- Add a flag to `Locale` called `undefinedAsRange` and a method called `toRange()` which returns a new `Locale` instance based on the one the method was called on. `match` would inspect the value of the flag for both `this` and `other` and act accordingly.

Now, I understand that this code is based on the existing C++/Rust code. With that in mind, I think this can be moved to a follow-up. Please file it if you agree.
